sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala

```diff
@@ -350,7 +350,14 @@ object DataType {
   }
 
   private def stringTypeWithCollation(collationName: String): StringType = {
-    StringType(CollationFactory.collationNameToId(collationName))
+    try {
+      StringType(CollationFactory.collationNameToId(collationName))
+    }
+    catch {
+      case e: SparkException if e.getCondition == "COLLATION_INVALID_NAME" &&
```

Comment:
I'm not really a fan of catching exceptions like this and comparing its string values; can we maybe create a small wrapper around `SparkException` with something like `InvalidCollationNameException`?